infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,293 @@
+
+        /**
+         * Get the total number of addresses in range.
+         * {@link #endAddress} and {@link #startAddress} can only be non-negative longs such that
+         * {@link #endAddress} >= {@link #startAddress}.
+         *
+         * @return Total number of addresses in this segment.
+         */
+        public long getTotal() {
+            return endAddress - startAddress + 1L;
+        }
+
+        public static class TransferSegmentBuilder {
+
+            public void verify() {
+                if (startAddress < 0L || endAddress < 0L) {
+                    throw new IllegalStateException(
+                            String.format("Start: %s or end: %s " +
+                                    "can not be negative.", startAddress, endAddress));
+                }
+                if (startAddress > endAddress) {
+                    throw new IllegalStateException(
+                            String.format("Start: %s can not be " +
+                                    "greater than end: %s.", startAddress, endAddress));
+                }
+
+                if (status == null) {
+                    throw new IllegalStateException("Status should be defined.");
+                }
+            }
+
+            public TransferSegment build() {
+                verify();
+                return new TransferSegment(startAddress, endAddress, status);
+            }
+        }
+    }
+
+    /**
+     * A data class that represents a status of a segment to be transferred.
+     */
+    @Getter
+    @ToString
+    @EqualsAndHashCode
+    @Builder
+    public static class TransferSegmentStatus {
+        /**
+         * States of the segment:
+         * - {@link SegmentState#NOT_TRANSFERRED}: Segment is not transferred.
+         * - {@link SegmentState#TRANSFERRED}: Segment was transferred fully.
+         * - {@link SegmentState#RESTORED}: Segment was restored, and is present in the current layout.
+         * - {@link SegmentState#FAILED}: The state transfer of a segment has failed.
+         */
+        public enum SegmentState {
+            NOT_TRANSFERRED,
+            TRANSFERRED,
+            RESTORED,
+            FAILED
+        }
+
+        /**
+         * A state of a segment.
+         */
+        @Default
+        private final SegmentState segmentState = NOT_TRANSFERRED;
+        /**
+         * Total number of records transferred for this segment.
+         */
+        @Default
+        private final long totalTransferred = 0L;
+        /**
+         * An optional cause of failure for this segment.
+         */
+        @Default
+        @Exclude
+        private final Optional<TransferSegmentException> causeOfFailure = Optional.empty();
+    }
+
+    /**
+     * A log unit client to the current node.
+     */
+    @Getter
+    @NonNull
+    private final LogUnitClient logUnitClient;
+
+    /**
+     * A size of one batch of transfer.
+     */
+    @Getter
+    @NonNull
+    private final int batchSize;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651453)